src/client/src/client/client.rs

```diff
@@ -37,15 +38,15 @@ pub(crate) trait SerialFrame: Sync + Send + Display {
     }
 
     /// send and receive
-    async fn send_receive<R>(&mut self, request: R) -> Result<R::Response, FlvSocketError>
+    async fn send_receive<R>(&self, request: R) -> Result<R::Response, FlvSocketError>
     where
         R: Request + Send + Sync;
 }
 
 /// This sockets knows about support versions
 /// Version information are automatically  insert into request
 pub struct VersionedSocket {
-    socket: AllFlvSocket,
+    socket: Mutex<AllFlvSocket>,
```

Comment:
I have prototype branch: https://github.com/sehz/fluvio/tree/duplicate_socket which try to use thread safe TcpStream.  In that case, you can duplicate sockets without problem.